src/index/CompressedRelation.cpp

```diff
@@ -1599,15 +1659,31 @@ auto CompressedRelationWriter::createPermutationPair(
     for (size_t idx : ad_utility::integerRange(block.numRows())) {
       Id col0Id = firstCol[idx];
       decltype(auto) curRemainingCols = permutedCols[idx];
+
       if (col0Id != col0IdCurrentRelation) {
         finishRelation();
         col0IdCurrentRelation = col0Id;
       }
+
+      // Check if we need to create a new block before adding the current
+      // triple. We create a new block if:
+      // 1. The relation buffer is at the block size limit, AND
+      // 2. The current triple has different first three columns than the last
+      //    triple in the buffer (to ensure equal triples stay in same block)
+      if (relation.size() >= blocksize && !relation.empty()) {
+        // Compare first three columns of current triple with last buffered
+        // triple
+        const auto& lastBufferedRow = relation.back();
+        if (curRemainingCols[0] != lastBufferedRow[0] ||
+            curRemainingCols[1] != lastBufferedRow[1] ||
+            curRemainingCols[2] != lastBufferedRow[2]) {
+          addBlockForLargeRelation();
+        }
+      }
+
       distinctCol1Counter(curRemainingCols[c1Idx]);
       relation.push_back(curRemainingCols);
-      if (relation.size() >= blocksize) {
-        addBlockForLargeRelation();
-      }
```

Comment:
Why is this code moved up, or rather, why was it down here previously? Also, how does this code relate to `addCompleteLargeRelation`?

Reply:
Previously, we just started new blocks whenever we felt like it.
This doesn't work anymore, because we have to make sure that the duplicates stay in the same block.